java-support/src/test/scala/io/cloudstate/javasupport/impl/eventsourced/EventSourcedImplSpec.scala

```diff
@@ -0,0 +1,211 @@
+      entity.send(command(1, "cart", "GetCart"))
+      entity.expect(reply(1, EmptyCart))
+      entity.send(command(2, "cart", "AddItem", addItem("abc", "apple", 1)))
+      entity.expect(reply(2, EmptyPayload, itemAdded("abc", "apple", 1)))
+      entity.send(command(3, "cart", "AddItem", addItem("abc", "apple", 2)))
+      entity.expect(reply(3, EmptyPayload, Seq(itemAdded("abc", "apple", 2)), cartSnapshot(Item("abc", "apple", 3))))
+      entity.send(command(4, "cart", "GetCart"))
+      entity.expect(reply(4, cart(Item("abc", "apple", 3))))
+      entity.send(command(5, "cart", "AddItem", addItem("123", "banana", 4)))
+      entity.expect(reply(5, EmptyPayload, itemAdded("123", "banana", 4)))
+      entity.passivate()
+      val reactivated = client.connect
+      reactivated.send(init(ShoppingCart.Name, "cart", snapshot(3, cartSnapshot(Item("abc", "apple", 3)))))
+      reactivated.send(event(4, itemAdded("123", "banana", 4)))
+      reactivated.send(command(1, "cart", "GetCart"))
+      reactivated.expect(reply(1, cart(Item("abc", "apple", 3), Item("123", "banana", 4))))
+      reactivated.passivate()
+    }
+
+    "error when first message is not init" in {
+      val entity = client.connect
+      entity.send(command(1, "cart", "command"))
+      entity.expectError("INTERNAL: Cloudstate protocol error: Expected Init message")
+    }
+
+    "error when service doesn't exist" in {
+      val entity = client.connect
+      entity.send(init(serviceName = "DoesNotExist", entityId = "cart"))
+      entity.expectError("INTERNAL: Cloudstate protocol error: Service not found: DoesNotExist")
+    }
+
+    "error when command payload is missing" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(command(1, "cart", "command", payload = None))
+      entity.expectError("INTERNAL: Cloudstate protocol error: No command payload")
+    }
+
+    "error when command entity id is incorrect" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart1"))
+      entity.send(command(1, "cart2", "command"))
+      entity.expectError(
+        "INTERNAL: Cloudstate protocol error: Receiving entity is not the intended recipient of command"
+      )
+    }
+
+    "fail when command handler does not exist" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(command(1, "cart", "foo"))
+      val expectedMessage = s"No command handler found for command [foo] on class ${ShoppingCart.TestCartClassName}"
+      entity.expect(failure(1, s"Unexpected failure: $expectedMessage"))
+      entity.passivate()
+    }
+
+    "fail when command uses context fail" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(command(1, "cart", "AddItem", addItem("foo", "bar", -1)))
+      entity.expect(failure(1, "Cannot add negative quantity of item [foo]"))
```

Comment:
the current TCK still mentions:
>// FIXME Expects entity.Failure, but gets lientAction.Action.Failure(Failure(commandId, msg)))

here: https://github.com/cloudstateio/cloudstate/blob/v0.5.1/tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala#L230

And I'm curious if we should expect an `actionFailure` for this kind of error, where the user function signals an error through `context.Fail()`.

I remember the TCK being inconsistent in this regard I think and the current Go implementation still sends a `client action failure`, even if a failure would be appropriate.

Reply:
Good point. Which level to return failures should be looked at more closely — with either `EventSourcedStreamOut.Message.Failure` and `ClientAction.Action.Failure`.

I was initially looking to just make sure errors are reported, rather than the EOF in the current behaviour. But it would make sense to make sure the protocol is sound now too.

`Context.fail` has the same behaviour as before — returns a client action failure. The testkit messages currently always assume a client action failure — I'll update to distinguish these failures.

For the unexpected exceptions in handlers it was easy to have it also be a context error and report in the same way, with a client action failure — this should probably be an event sourced message failure instead. But it seems that the current behaviour of the proxy, however, is to translate these failures into client action failures anyway, and continuing with processing rather than cancelling the stream.

https://github.com/cloudstateio/cloudstate/blob/master/proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedEntity.scala#L332-L337

https://github.com/cloudstateio/cloudstate/blob/master/proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedEntity.scala#L266-L269

If any other exceptions are thrown (outside of the handler, for protocol errors, service not found, wrong recipient, or so on), I've updated this to throw a gRPC exception when the stream is stopping, so that it gets reported while cancelling — but this could also be an explicit event sourced message failure instead of a stream error, and there's a previous fixme suggesting this.

So currently, the event sourced message failure is not expected at all, it's either a client action failure or it's the stream completing with an error. At least in the java support. And if event sourced message failures are returned, they are not currently treated differently in the proxy, but translated into client action failures.

Thanks for raising this @marcellanz. I see you've also asked about failures in the draft spec. I can look at updating things to be in line with the spec discussions — where client action failure is only used to signal an explicit context.fail, and event sourced message failures are used for all other exceptions and these are treated as fatal on the proxy side.

Reply:
I've updated this PR to be aligned with discussions — java-support only uses client action failure for explicit context.fail and all other exceptions are reported as entity failures, and the proxy now treats entity failures as fatal and doesn't leak specific details of these failures to the client.

Reply:
Thanks for looking at updating things to be in line with the spec discussions @pvlugter.

Probably out of the scope of this issue/PR comment, I'm interested in to define what is expected from a user entity after issuing fatal and non-fatal failures in the user entity now handled by the proxy differently in terms of expectations of state consistency of an entities local state.

This would not change anything changed with this PR I think, so I'll add my question to the draft spec.

Reply:
On the behaviour and currently implemented expectations:

* Action failures in replies are handled by the proxy in the same way as non-failure replies (same as it was before this PR). So the user entity is expected to be in a consistent state after explicit calls to `context.fail` — it can be interpreted as reply with failure message, and usually the state is unchanged (no events).

* Entity failures now error and stop the proxy entity actor and close the stream (this has changed behaviour in this PR, previously it continued processing). The user entity should be cleared and the stream closed for these unexpected failures.

I'm assuming these are aligned with the original intent and spec discussions now.

But what I'm hearing in your questions: should any failure, including `context.fail`, be considered fatal for the current state of the entity. So that a user can call `context.fail` anywhere that they think things have become inconsistent. I can see how that makes sense, especially with it being a `fail` call. In which case, I think we would want a different mechanism for returning status codes and descriptions for client request errors. Currently `context.fail` means reply with a client request error (immediately) but continue with this active entity otherwise, and to fail the entity fully an exception can be thrown.

Reply:
The fact is not very clear if the user wants to interrupt the flow due to a fatal error or just inform that an expectation has failed and he is only intending to inform the client.
Perhaps the context, as well as the protocol, allows to choose the type of failure or the desired return, perhaps this can be implemented with the two types of failure already existing. In Rust, you can wait when we want to report that an expectation has not been met, perhaps the context can prove the same type of mechanism, namely context.fail () for a catastrophic failure and context.expect () to just report, the first one finishes the flow trying to keep the state consistent while the second one follows the normal flow without closing it, probably both without events

Reply:
Exactly, this is what I thought about.

As soon as a command handler emits events, and the events get applied immediately to the local state, a later call to `context.fail` would leave the entities local state in an inconsistent state the proxy is not aware of I think. The entity could send the emitted events back to the proxy, `EventSourcedReply` would allow this, but this would leave the entity in an awkward semi-complete state I think. The option then could be, that the user entity informs the proxy about the `context.fail`-error with an action failure, but also lets the proxy re-establish its state; the proxy in this situation knows best about the entities consistent state.

Saying this, if the client sends the same command, the user entity will probably end in the same situation and again will fail as described with the context failed. But if the client sends a command with different input, because the client was informed through the client action failure, the entity could proceed.

Reply:
@marcellanz What you describe sounds like idempotency. That would be very desirable. It seems to me that context.fail would be called in commands never in events, which alone would guarantee that the state would be consistent, since events have not yet been issued.

Reply:
>It seems to me that context.fail would be called in commands never in events

Correct, should. But the following can happen:

- start to handle command
- **`call context.emit()`** which emit events and get them **immediately applied** as by current documentation [described](https://cloudstate.io/docs/core/current/user/lang/java/eventsourced.html#emitting-events) or see below 
(note: this is not done as documented in the JS implementation currently, see https://github.com/cloudstateio/cloudstate/issues/375)
- **`call context.fail()`**
- the user even can call again context.emit(), but not `context.fail()`

We even say:
>this both validates that the event can be applied to the current state, as well as updates the state so that subsequent processing in the command handler can use it.

so this is intended to happen.


**emitting-events**
https://cloudstate.io/docs/core/current/user/lang/java/eventsourced.html#emitting-events states:
**Invoking emit will immediately invoke the associated event handler** for that event - this both validates that the event can be applied to the current state, as well as updates the state so that subsequent processing in the command handler can use it.

Reply:
As we say we'd "**validate that the event can be applied to the current state**", implies that we, if we fail, are inconsistent in our local state. If we don't allow `context.fail` to be called after `context.emit()` every failure in `context.emit()` or exception after `context.emit()`, would lead to a fatal, stream closing, failure.

Reply:
Yes, I agree, but this is because in the current scenario it seems to be assumed that this is not really a real failure. I think that one should first separate the concepts of failure and validation and then define the best chain of events to be carried out.
For me it would be natural to call context.emit (), context.expect (), context.emit () because I am just reporting something, but against intuitive to call context.emit (), context.fail (), context.emit because it seems that fail now should break the whole thing and interrupt the flow of events until my last issue () before fail

Reply:
I see. I think this is quite a burden for the user to ensure doing the right thing at the right time.

There is something else, that we describe in [terminology](https://cloudstate.io/docs/core/current/user/features/eventsourced.html) – see below.

During command handling, we are not allowed to update the local state, but we also describe that we do exactly this, because we interweave the span of the command handling with the update of the local state right within `context.emit`.

One understanding the relation of commands and events, events are side effects happened after a command has completed as a whole I think. If the command did fail, there should be no events, and there are none after a `context.fail` because the proxy does not know about them. A further command would emit, based on inconsistent local state, inconsistent new events and if this new command succeeds, the proxy would get tainted events from the entity I think.

**Command handler**
>A command handler is the code that handles a command. It may validate the command using the current state, and may emit events as part of its processing. A command handler **must not update the state of the entity directly**, only indirectly by emitting events. If a command handler does update the state, then when the entity is passivated (removed from memory), those updates will be lost.

Reply:
I see this might all be, we should assume it does not happen, but if it does, it corrupts state and is not easily detected. 

The Client Action Failure is just a normal hint for the client something did not go well, like the negative added products in the shopping cart, it might be just a hint for the user to "try" again and is not exceptional. Especially in this case, if the user retries, failed validation in the command should not lead to corrupted state.

Reply:
There are flaws and flaws in how the user informs about proxy or that each one must define the result of the flow of events.

Reply:
Agree, we need to define this clearly. And make it as intuitive as possible for users.

With the scenario of emit and then fail, the java-support implementation will actually have entities in inconsistent states. Events are applied immediately on emit, as described, but no events are returned in the reply for action failures. So after an emit-then-fail a subsequent get on the entity will return an updated state, but once the entity is passivated and reactivated the state will not have those events. Users can't call emit after fail though (as fail throws an exception in the java-support and node-support implementations). So the current implementation does have the possible bug suggested by failing after immediately applied emits. It needs to either respond with any events emitted and applied before a fail so these can be persisted, or wait with applying events until after successful handling of a command (like the node-support implementation does).

It's useful to have the emitted events validated against the event handlers, before responding, but we need to avoid this being in conflict with a likely expectation for handlers to be atomic — all events or none applied — and that `fail` is probably understood as failing the whole handler, including any events emitted. The usual approach is to wait for events to be persisted successfully before updating the in-memory state, but here that would require additional communication from the proxy, but the entity could still have atomicity for handlers by applying after command processing. So I'm in favour of the node-support version being the expected behaviour.

I'm not sure that we would need multiple ways to fail, as users should only need to fail an invalid command, and not the entity as a whole. Only valid commands should be applied, with the state kept consistent. The current `context.fail` as meaning fail the current command invocation, with no events applied, makes sense to me.

I've extended this test to show the incorrect behaviour. And I think the change we should make now is having the java-support delay applying events until after successful command processing.

Reply:
Thanks @pvlugter for the summary, I agree.

Reply:
An additonal thought:
The entity actually knows how to reset its local state to the state it had before the command handler started through a snapshot it would take and apply to itself, or hold the last snapshot plus events received from its initial state. This would prevent any communication from the proxy to restore its local state.

Reply:
@marcellanz  thinks this is complicated, it would solve the problem but it breaks the basic premise that it is the proxy that deals with the state. Guys, I know that we talk a lot about support for Java and Node, but as the other languages are, maybe the others are also inconsistent with the behavior that Peter considers expected. Is there a way for the proxy to fully accomplish the task?
@pvlugter  I don't like to keep just context.fail () semantically for me it's like I'm returning an HTTP 500 to the client, when in fact my intention would be an HTTP 400 Bad Request

Reply:
I think it's not. I don't say to do it that way but that it's a possibility and could be considered.
The sequence of how entities get their state is well defined, sequential and it would not be different from any snapshot or event the entity would apply instructed by the proxy right before it gets any new events or a command from the proxy. Also, and especially with the event sourcing procotol, the proxy does not know what is inside a snapshot or an event, it could be anything. The entity in this regard has to manage its state always in a consistent manner. So just food for thoughts.

I agree not to implement anything complicated, the simpler the better.

Reply:
@sleipnir the current Go support is implemented by the spec and docs and has the described flaw right now.

Reply:
Hmm ... I don't know, for me the premise breaks and the premise breaks so to implement this we must change the premise, "proxy manages the state of the application", "we move to the proxy shares the state with the application", I definitely don't know, I would have to think about it a lot to have a plausible answer and I have a lot of headaches at the moment. :(

Reply:
@sleipnir I hope not because of the idea shared by me before. I'm pretty sure we might not are aligned or this is a misunderstanding. What I wrote about using snapshots to revert on a `context.fail` is an optimization of what Peter has described with:

>The usual approach is to wait for events to be persisted successfully before updating the in-memory state, but here **that would require additional communication from the proxy**, but the entity could still have atomicity for handlers by applying after command processing.

where events would be applied after the command and its events got back to the proxy and the proxy would send those events to the user entity to get applied, or I think, the proxy would re-establish the state of the entity after a `context.fail` ed entity error.

I'm curious if we have multiple talking points here, one of
 **a)** what semantics does `context.fail` have regarding the local state, and
 **b)** how should the emitted events get applied, during or after the command succeeded?

I think for **b)** we have a relevant aspect of what is described why to have the local state changed during the time the command handler is running. One is to validate the event can be applied and the other is to "_updates the state so that subsequent processing in the command handler can use it_" defined in the documentation:

>Invoking emit will immediately invoke the associated event handler for that event - this **both validates that the event can be applied to the current state, as well as updates the state so that subsequent processing in the command handler can use it**.

If events are not applied to local state, it could be difficult to a command handler go get the right values of the local state after the `context.emit` applied the to the local state.

Perhaps we could discuss these issues and ideas on the next contributor call :)

Reply:
Yes @marcellanz , it would be nice to discuss this in a collaborator call, if you participate in any please update the minutes of the meeting, I can rarely participate but I like to read about what was discussed at the meeting.

Reply:
Updated java-support to only apply events after successful command processing. This still validates and applies the emitted events; if any of them fail then it will crash the entity before the reply is sent.

I think it works to have `context.fail` mean fail the current command (without applying or persisting any events). This should only be available to command contexts.

The protocol does allow returning an action failure along with events, snapshots, or effects — which will be applied by the proxy. If this should never be the case for language supports, then we may want to adjust the protocol so this is not supported.

Reply:
Thanks. I'll implement the same for Go Support.

Reply:
@sleipnir we might have to check to have the same semantics with other language support libraries.

Reply:
@pvlugter thanks for adapting the java support. i will updated for other event-sourced based model like CRUD.

Reply:
Thanks @pvlugter 
@marcellanz @pvlugter I will check out Kotlin, Dart, C#, Springboot I just need to update the version of the java-support library, so it would be interesting to release a release of java-support with this fix (alias it would be nice to release a release with everything that has been done since the last version ), perhaps this is the same case for Kotlin. I'm going to take a look at Python too, but if you want to see that for me that's fine. I think that would be it for now, if anyone remembers somewhere else to look let me know.

Reply:
thanks @sleipnir, I can check the python-support.

Reply:
@marcellanz really great for finding the bug!

Reply:
I can publish another 0.5.x release for java-support with the failure handling changes. I was originally going to wait for the metadata and projection changes to be finished before next releases, but no problem to have extra releases.

Reply:
@sleipnir The Python Support does apply events after the command handler ran, `context.emit` just collects the event:
https://github.com/cloudstateio/python-support/blob/v0.1.1/cloudstate/evensourced_servicer.py#L100-L103

no change is needed here IMO.

Reply:
Very good

Reply:
java-support 0.5.2 is published to bintray and synced to maven central

Reply:
@pvlugter I tried to run the TCK matrix and the docker release images where not available for 0.5.2. I think the release travis build [did not run through](https://travis-ci.com/github/cloudstateio/cloudstate/jobs/368079957#L496).

Reply:
I've only released java-support 0.5.2. The versions are similar, but proxy (and TCK), java-support, and node-support are on separate release cycles. The `latest` version of the docker images will have the changes. I didn't release cloudstate core, as the metadata and projection work is in progress and eventing has been disabled as part of this. Eventing may not be used anywhere, so can certainly release if it's needed. 

On the travis build failure: since it's the first time using the new publish via bintray, the maven central sync still needed to be set up — it can't be enabled until something is published first. Bintray have approved the sync now, and subsequent releases will work fine.